source/extensions/transport_sockets/tls/ocsp/asn1_utility.h

```diff
@@ -0,0 +1,208 @@
+ * This relies heavily on the 'openssl/bytestring' API which
+ * is BoringSSL's recommended interface for parsing DER-encoded
+ * ASN.1 data when there is not an existing wrapper.
+ * This is not a complete library for ASN.1 parsing and primarily
+ * serves as abstractions for the OCSP module, but can be
+ * extended and moved into a general utility to support parsing of
+ * additional ASN.1 objects.
+ *
+ * Each function adheres to the invariant that given a reference
+ * to a crypto bytestring (CBS&), it will parse the specified
+ * ASN.1 element and advance `cbs` over it.
+ *
+ * An exception is thrown if the bytestring is malformed or does
+ * not match the specified ASN.1 object. The position
+ * of `cbs` is not reliable after an exception is thrown.
+ */
+class Asn1Utility {
+public:
+  ~Asn1Utility() = default;
+
+  /**
+   * Extracts the full contents of `cbs` as a string.
+   * This copies the data in `cbs`.
+   *
+   * @param cbs a CBS& that refers to the current document position
+   * @returns std::string containing the contents of `cbs`
+   */
+  static std::string cbsToString(CBS& cbs);
+
+  /**
+   * Parses all elements of an ASN.1 SEQUENCE OF. |parse_element| must
+   * advance its input CBS& over the entire element.
+   *
+   * @param cbs a CBS& that refers to an ASN.1 SEQUENCE OF object
+   * @param parse_element an Asn1ParsingFunc<T> used to parse each element
+   * @returns std::vector<T> containing the parsed elements of the sequence.
+   * @throws Envoy::EnvoyException if `cbs` does not point to a well-formed
+   * SEQUENCE OF object.
+   */
+  template <typename T>
+  static std::vector<T> parseSequenceOf(CBS& cbs, Asn1ParsingFunc<T> parse_element);
+
+  /**
+   * Checks if an explicitly tagged optional element of |tag| is present and
+   * if so parses its value with |parse_data|. If the element is not present,
+   * `cbs` is not advanced.
+   *
+   * @param cbs a CBS& that refers to the current document position
+   * @param parse_data an Asn1ParsingFunc<T> used to parse the data if present
+   * @return absl::optional<T> with a T if `cbs` is of the specified tag,
+   * else nullopt
+   */
+  template <typename T>
+  static absl::optional<T> parseOptional(CBS& cbs, Asn1ParsingFunc<T> parse_data, unsigned tag);
+
+  /**
+   * Returns whether or not an element explicitly tagged with |tag| is present
+   * at `cbs`. If so, `cbs` is advanced over the optional and assigns
+   * |data| to the inner element, if |data| is not nullptr.
+   * If `cbs` does not contain |tag|, `cbs` remains at the same position.
+   *
+   * @param cbs a CBS& that refers to the current document position
+   * @param data a CBS& that is set to the contents of `cbs`
+   * @param an explicit tag indicating an optional value
+   *
+   * @returns bool whether `cbs` points to an element tagged with |tag|
+   * @throws Envoy::EnvoyException if `cbs` is a malformed TLV bytestring
+   */
+  static bool isOptionalPresent(CBS& cbs, CBS* data, unsigned tag);
```

Comment:
I found this name confusing when reading code below, because it wasn't clear that this advances cbs. Maybe it should be `getOptional()`?